torch_geometric/nn/conv/edp_conv.py

```diff
@@ -0,0 +1,183 @@
+from typing import Callable, List, Union
+
+import torch
+from torch import Tensor
+from torch.nn import Module
+
+from torch_geometric.nn.conv import MessagePassing
+from torch_geometric.typing import Adj, OptPairTensor, Size
+from torch_geometric.utils import dense_to_sparse, to_dense_adj
+
+
+class MultiChannelConv(MessagePassing):
+    r"""The multi channel convolutional operator from the `"Permutation
+    Invariant Graph Generation via Score-Based Generative Modeling"
+    <https://arxiv.org/abs/2003.00638>`_ paper.
+    "The intuition is to run message-passing simultaneously on many different
+    graphs, and collect the node features from all the channels via
+    concatenation."
+    The key thing is edhes have weights, so we have weighted messages
+    to aggregate.
+
+    .. math::
+        \widetilde{X}^{\prime}_{[c,\cdot]}=A_{[c,\cdot]}X_{[\cdot]},
+        for c = 0, 1, ..., C-1
+
+        \mathbf{x}^{\prime}_i = \texttt{MLP}_{\texttt{Node}} \left(
+        \texttt{CONCAT}(\widetilde{X}_{[c,i]} + (1 + \epsilon)X_i\ |
+        \ c = 0, 1, \cdots, C-1 )  \right)
+
+    here MLP denotes a neural network, in original version it is an MLP.
+
+    Args:
+        nn (torch.nn.Module): A neural network :math:`\text{MLP}_{\text{Node}}`
+            that maps node features :obj:`x` of shape :obj:`[-1, in_channels]`
+            to shape :obj:`[-1, out_channels]`, *e.g.*, defined by
+            :class:`torch.nn.Sequential`.
+        eps (float, optional): (Initial) :math:`\epsilon`-value.
+            (default: :obj:`0.`)
+        train_eps (bool, optional): If set to :obj:`True`, :math:`\epsilon`
+            will be a trainable parameter. (default: :obj:`False`)
+        **kwargs (optional): Additional arguments of
+            :class:`torch_geometric.nn.conv.MessagePassing`.
+
+    Shapes:
+        - **input:**
+          node features :math:`(|\mathcal{V}|, F_{in})`,
+          edge indices :math:`(2, |\mathcal{E}|)` for each element of List of
+          :math:`C` channels
+          edge weights :math:`(\mathcal{E}|)` for each element of List of
+          :math:`C` channels
+        - **output:** node features :math:`(|\mathcal{V}|, F_{out})`
+    """
+    def __init__(self, nn: Callable, eps: float = 0., train_eps: bool = False,
+                 **kwargs):
+        kwargs.setdefault('aggr', 'add')
+        super().__init__(**kwargs)
+        self.nn = nn
+        self.initial_eps = eps
+        if train_eps:
+            self.eps = torch.nn.Parameter(torch.empty(1))
+        else:
+            self.register_buffer('eps', torch.empty(1))
+
+    def message(self, x_j, edge_weight):
```

Comment:
@atimashov should we add docstrings for all the methods here?

Reply:
@Loquats I checked a few implementations of layers, folks don't add. But theoretically it might be beneficial.